### ISSUE TITLE
Explore offline generation of sequences

### ISSUE THREAD

**Issue by fmpaezri** [jccaicedo/localization-agent#25]

**Comment by fmpaezri** [comment]
I generated a bunch of simulated sequences and dumped them to videos. This takes ~<1s per video and the mean video size is 200KB. This is a lot less than storing the images (raw bytes, JPEG compression might reduce it but OOT dataset takes 1,4G in images and 110M in videos for example)

60 frames x 320 width x 240 height x 3 channels x 1 byte = 13824000 bytes

but the video generation dumps the images to a temporal folder and then calls a process to encode them, but it seems to be done in parallel, as CPU usage was 300% and measuring time with %time returns

CPU times: user 814 ms, sys: 1.34 s, total: 2.15 s
Wall time: 837 ms

TODO: measure video reading time

Reading the video creates a dependency on OpenCV.

I tried generating an HDF5 file of a batch of activations also. This takes more space than the videos but less than the images (442368000 bytes for 32 sequences)

32 batch size x 60 frames x 7 width x 7 height x 1024 channels/filters x 4 bytes = 385351680 bytes

but saving the file took 2s, did not tried with a pickle file. Reading time is negligible, 4ms.

The GPU is the bootleneck for feature extraction, we could had a lot of machines generating videos and sending them to a central GPU node to at least keep it almost always busy. It would be better if there were also many GPU nodes sending features to a training node. But i dont know if the overhead is worth, specially when saving the activations. It would also be possible to avoid saving to disk by sending arrays over the network, but the transport overhead is still unknown.

I changed the parallel simulation and now the simulation and feature extraction seem to be more simultaneous. With a batch size of 128 and a GPU batch of 4 this saved ~10s, passing from ~40 to ~30s. The 40s are ~23s for simulation, ~15s forwarding and ~2s training, and the 30s are ~6s of exclusive simulation, ~24s of simultaneous extraction and simulation and ~2s training. I used 128 for batch size as this took ~9-12GB RAM when collecting frames and features (it could be more if a lot or RAM is available). I am not sure why feature extraction passed from ~15s alone to ~24s with simulation, as they are executed in CPU and GPU independently. Maybe because the process pool uses all cores by default so the main process has to also compete for CPU time?
